internal/backends/mcu/pico_st7789.rs

```diff
@@ -153,16 +185,89 @@ pub fn init() {
         vec![Rgb565::default(); DISPLAY_SIZE.width as _].leak(),
         stolen_spi,
     );
-
-    crate::init_with_display(PicoDevices {
+    let mut buffer_provider = DrawBuffer {
         display,
-        touch,
-        last_touch: Default::default(),
-        timer,
         buffer: vec![Rgb565::default(); DISPLAY_SIZE.width as _].leak(),
         pio: Some(pio),
         stolen_pin: (dc_copy, cs_copy),
-    });
+        dirty_region: Default::default(),
+    };
+
+    let mut last_touch = None;
+    loop {
+        i_slint_core::timers::update_timers();
+        if NEEDS_REDRAW.load(Ordering::Relaxed) {
+            NEEDS_REDRAW.store(false, Ordering::Relaxed);
+            if let Some(window) = WINDOW.with(|x| x.borrow().clone()) {
+                let runtime_window = window.self_weak.upgrade().unwrap();
+                runtime_window.update_window_properties();
+
+                let scale_factor = runtime_window.scale_factor();
+                let size = DISPLAY_SIZE.to_f32() / scale_factor;
+                runtime_window.set_window_item_geometry(size.width as _, size.height as _);
+
+                window.renderer.render_by_line(&runtime_window.into(), &mut buffer_provider);
+                buffer_provider.flush_frame();
+            }
+        }
+
+        // handle touch event
+        let button = i_slint_core::items::PointerEventButton::Left;
+        if let Some(mut event) = touch
+            .read()
+            .map_err(|_| ())
+            .unwrap()
+            .map(|point| {
+                let size = DISPLAY_SIZE.to_f32();
+                let pos = euclid::point2(point.x * size.width, point.y * size.height).cast();
+                match last_touch.replace(pos) {
+                    Some(_) => i_slint_core::input::MouseEvent::MouseMoved { pos },
+                    None => i_slint_core::input::MouseEvent::MousePressed { pos, button },
+                }
+            })
+            .or_else(|| {
+                last_touch
+                    .take()
+                    .map(|pos| i_slint_core::input::MouseEvent::MouseReleased { pos, button })
+            })
+        {
+            if let Some(window) = WINDOW.with(|x| x.borrow().clone()) {
+                let w = window.self_weak.upgrade().unwrap();
+                // scale the event by the scale factor:
+                if let Some(p) = event.pos() {
+                    event.translate((p.cast() / w.scale_factor()).cast() - p);
+                }
+                w.process_mouse_input(event);
+            }
+        } else if i_slint_core::animations::CURRENT_ANIMATION_DRIVER
+            .with(|driver| !driver.has_active_animations())
+        {
```

Comment:
I guess this is the other place where internal API is still used?

Reply:
Yes, there is still quite a bit of private API in use